File: 02/cpp/a.cpp
#include "util.h"

long long solve(vector<string>& lines) {
  long long result = 0;

  for (const string& line : lines) {
    const vector<string> parts = split(line, ": ");

    bool success = true;
    for (const string& item1 : split(parts[1], "; ")) {
      int red = 0, green = 0, blue = 0;
      for (const string& item2 : split(item1, ", ")) {
        const vector<string>& t4 = split(item2, " ");
        if ("red" == t4[1]) {
          red += stoi(t4[0]);
        } else if ("green" == t4[1]) {
          green += stoi(t4[0]);
        } else if ("blue" == t4[1]) {
          blue += stoi(t4[0]);
        }
      }
      if (12 < red || 13 < green || 14 < blue) {
        success = false;
      }
    }

    const int gameNum = stoi(parts[0].substr(string("Game ").size()));
    if (success) {
      result += gameNum;
    }
  }

  return result;
}

File: 02/rust/src/a.rs
pub fn solve(lines: &[String]) -> i32 {
  let mut result = 0;
  for line in lines {
    let parts: Vec<&str> = line.split(": ").collect();
    assert!(2 == parts.len());

    let mut success = true;
    for item1 in parts[1].split("; ") {
      let mut red = 0;
      let mut green = 0;
      let mut blue = 0;
      for item2 in item1.split(", ") {
        let t4: Vec<&str> = item2.split(" ").collect();
        let num1: i32 = t4[0].parse().unwrap();
        if "red" == t4[1] {
          red += num1;
        } else if "green" == t4[1] {
          green += num1;
        } else if "blue" == t4[1] {
          blue += num1;
        }
      }
      if 12 < red || 13 < green || 14 < blue {
        success = false;
      }
    }
    let game_num: i32 = parts[0].get(("Game ".len())..).unwrap().parse().unwrap();
    if success {
      result += game_num;
    }
  }
  return result;
}
